tools/packager/main.go

```diff
@@ -15,50 +18,88 @@ var (
 	defer f.Close()
+
 	zw := gzip.NewWriter(f)
 	defer zw.Close()
+
 	tw := archive.NewNormalizingTarWriter(tar.NewWriter(zw))
 	tw.WithUID(0)
 	tw.WithGID(0)
 	defer tw.Close()
 
-	descriptorInfo, err := os.Stat(descriptorPath)
+	templateContents, err := ioutil.ReadFile(descriptorPath)
+	handle(err, fmt.Sprintf("Failed to read descriptor file %s: %s", descriptorPath, err))
+
+	descriptorContents, err := fillTemplate(templateContents, map[string]interface{}{"lifecycle_version": version})
+	handle(err, fmt.Sprintf("Failed to fill template: %s", err))
+
+	descriptorTemplateInfo, err := os.Stat(descriptorPath)
+	handle(err, fmt.Sprintf("Failed to stat descriptor template file %s: %s", descriptorPath, err))
+
+	tempDir, err := ioutil.TempDir("", "lifecycle-descriptor")
+	handle(err, fmt.Sprintf("Failed to create a temp directory: %s", err))
+
+	tempFile, err := os.Create(filepath.Join(tempDir, "lifecycle.toml"))
+	handle(err, fmt.Sprintf("Failed to create a temp file: %s", err))
+
+	err = ioutil.WriteFile(tempFile.Name(), descriptorContents, descriptorTemplateInfo.Mode())
+	handle(err, fmt.Sprintf("Failed to write descriptor contents to file %s: %s", tempFile.Name(), err))
+
+	err = os.Chdir(tempDir)
+	handle(err, fmt.Sprintf("Failed to switch directories to %s: %s", tempDir, err))
+
+	descriptorInfo, err := os.Stat(tempFile.Name())
+	handle(err, fmt.Sprintf("Failed to stat descriptor file %s: %s", tempFile.Name(), err))
+
+	err = archive.AddFileToArchive(tw, "lifecycle.toml", descriptorInfo)
+	handle(err, fmt.Sprintf("Failed to write descriptor to archive: %s", err))
+
+	err = os.Chdir(filepath.Dir(inputDir))
+	handle(err, fmt.Sprintf("Failed to switch directories to %s: %s", filepath.Dir(inputDir), err))
+
+	err = archive.AddDirToArchive(tw, filepath.Base(inputDir))
+	handle(err, fmt.Sprintf("Failed to write dir to archive: %s", err))
+}
+
+func fillTemplate(templateContents []byte, data map[string]interface{}) ([]byte, error) {
+	tpl, err := template.New("").Parse(string(templateContents))
 	if err != nil {
-		fmt.Printf("Failed stat descriptor file at path %s: %s", descriptorPath, err)
-		os.Exit(3)
-	}
-	if err := archive.AddFileToArchive(tw, "lifecycle.toml", descriptorInfo); err != nil {
-		fmt.Printf("Failed to write descriptor to archive: %s", err)
-		os.Exit(4)
+		return []byte{}, err
 	}
 
-	if err := os.Chdir(filepath.Dir(inputDir)); err != nil {
-		fmt.Printf("Failed to switch directories to %s: %s", filepath.Dir(inputDir), err)
-		os.Exit(5)
+	var templatedContent bytes.Buffer
+	err = tpl.Execute(&templatedContent, data)
+	if err != nil {
+		return []byte{}, err
 	}
 
-	if err := archive.AddDirToArchive(tw, filepath.Base(inputDir)); err != nil {
-		fmt.Printf("Failed to write dir to archive: %s", err)
-		os.Exit(6)
+	return templatedContent.Bytes(), nil
+}
+
+func handle(err error, msg string) {
+	if err != nil {
+		fmt.Println(msg)
+		os.Exit(exitCode)
 	}
+	exitCode++
 }
```

Comment:
It not important to preserve the different error codes here. I think the most natural way to reduce duplication in this file would be to pull the bulk of the logic into a function and return and error from that function. Then in `main`, if `err != nil` print the error and exit with a nonzero code.